src/openforms/forms/tests/e2e_tests/test_form_designer.py

```diff
@@ -2076,3 +2077,71 @@ def setUpTestData():
                 'duplicate-key: in "FORM DEFINITION #1" and "FORM DEFINITION #2"'
                 'duplicate-key-2: in "FORM DEFINITION #1" and "FORM DEFINITION #3"'
             )
+
+
+class FormDesignerJsonLogicTests(E2ETestCase):
+    async def test_variable_options(self) -> None:
+        @sync_to_async
+        def setUpTestData() -> Form:
+            return FormStepFactory.create(
+                form__name="Inferring type of variable options",
+                form_definition__name="Pick your poison",
+                form_definition__configuration={
+                    "components": [
+                        {"label": "Single choice", "type": "radio", "key": "single"},
+                        {
+                            "label": "Multiple choice",
+                            "type": "selectboxes",
+                            "key": "multi",
+                        },
+                        {"label": "Dropdown", "type": "select", "key": "dropdown"},
+                    ],
+                },
+            ).form
+
+        await create_superuser()
+        form = await setUpTestData()
+
+        edit_form_url = str(
+            furl(self.live_server_url)
+            / reverse("admin:forms_form_change", args=(form.pk,))
+        )
+
+        async def make_assertions():
+            await page.get_by_text("Keuzeopties", exact=True).click()
+            await page.get_by_text("Variabele", exact=True).click()
+            # clicking the "Items expression" label doesn't focus the editor :(
+            component = page.locator(
+                'css=[ref="component"]', has_text="Opties-expressie"
+            )
+            editor = component.locator("css=.ace_content")
+            await editor.click()
+            await editor.press("Control+a")
+            await editor.press("Backspace")
+            await editor.type('{"map": [{"var": "somevar"}]}')
+            await page.get_by_role("button", name="Opslaan").click()
+            await expect(page.locator("css=.formio-dialog-content")).to_be_visible()
+            await expect(
+                page.locator('css=[data-component-key="openForms.itemsExpression"]')
+            ).to_contain_text("Error")
+            await editor.click()
+            await editor.press("Control+a")
+            await editor.press("Backspace")
+            await editor.type(
+                '{"map": [{"var": "somevar"}, [{"var": ""}, {"cat": ["Optie ", {"var": ""}]}]]}'
+            )
```

Comment:
How does it know that `{"var": ""}` in the context of `{"var": "somevar"}` (is this checked that it is of type array?) is a string?

Reply:
I don't know the exact order, but it follows from 

 - `{"cat": ["Optie ", {"var": ""}]}`
 - the list of options is of `string[][]` / `list[list[str]]`, therefore the result of the `"map"` is also of that type.

It's why the `validateLogic` passed the expression into
```ts
const result = infer({'===': [logic, expected]}, {});
```
with `expected` = `[["", ""]]` and `logic` = (in this case) the `"map"`

It uses the fact that `"==="` is a `Callable[[T, T], boolean]`, so the parse of the result from the `infer` call is a lot easier. `infer({"===": [{"var": "a"}, {"var": "b"}]})` does *not* imply that `a` and `b` have the same value, just the same type.

`infer` shouldn't return a string, of course, but before changing it, I like to have more use cases.